### PR TITLE
Kalle/edurep json connection

### PR DESCRIPTION
**Pull request by kallepronk**
A few things that need feedback before this issue can be continued:
in the [extraction class](https://github.com/surfedushare/search-portal/blob/0e775459ed4f5ba174c86be586255adbb60a0b04/harvester/sources/extraction/edurep.py)

- On line 276 and 284 there are two keys with kind of the same name, are these different values or can one be removed?
- Where in the JSON input can information for the studies, ideas and learning_material_disciplines be found?
- There hasn't been done anything with an document id or general id. Is that missing somewhere?

These are the specific questions I have about the code. If there is anything else I'd welcome that info as well.

### CODE REVIEW

File: harvester/sources/extraction/edurep.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+from mimetypes import guess_type
+from hashlib import sha1
+from core.constants import HIGHER_EDUCATION_LEVELS
+from django.conf import settings
+from dateutil.parser import parse as date_parser
+from datagrowth.processors import ExtractProcessor
+from django.utils.text import slugify
+
+
+FILE_TYPE_TO_MIME_TYPE = {
```
Comment (fako) [inline harvester/sources/extraction/edurep.py]: This is problematic. We need to talk about this with Edurep if we really can't find the mime_type. Do they want us to look for ourselves?! This value is much more akin to our technical_type than it is to a mime_type.

File: harvester/sources/extraction/edurep.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            "title": node["schema:name"]["@value"],
+            "url": node["schema:url"],
+            "mime_type": FILE_TYPE_TO_MIME_TYPE.get(node["schema:encodingFormat"]),
+            "hash": sha1(node["schema:url"].encode("utf-8")).hexdigest(),
+            "copyright": node["schema:license"],
+            "access_rights": node["dcterms:accessRights"]
+        }
+
+    @classmethod
+    def get_language(cls, node):
```
Comment (fako) [inline harvester/sources/extraction/edurep.py]: You've adjusted this method, but the OBJECTIVE is not using this method. Take a close look at the OBJECTIVE yourself and correct where necessary.

File: harvester/sources/extraction/edurep.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            return "nl"
+        elif language == "English":
+            return "en"
+        return "unk"
+
+    @classmethod
+    def get_url(cls, node):
+        files = cls.get_files(node)
+        if not files:
+            return node["portalUrl"]
```
Comment (fako) [inline harvester/sources/extraction/edurep.py]: Maybe a node.get call with None as default is safer than an item access without any error handling. We'll find out if we run this on all 2000+ materials, but I'm expecting some crashes here. Let's buy some popcorn and watch :partying_face:

File: harvester/sources/extraction/edurep.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            license = "cc0"
+        else:
+            license = "cc-" + license
+        return slugify(f"{license}-{url_match.group('version')}")
+
+    @classmethod
+    def get_copyright(cls, node):
+        copyright = node["@type"]["lom:copyrightAndOtherRestrictions"]
+        if copyright is None:
+            return "yes"
```
Comment (fako) [inline harvester/sources/extraction/edurep.py]: This line now never looks at the copyright description. I think setting copyright to "yes" if it is None makes more sense than returning "yes", while description may hint at something more specific.

File: harvester/sources/extraction/edurep.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    def get_from_youtube(cls, node):
+        url = cls.get_url(node)
+        if not url:
+            return False
+        return cls.youtube_regex.match(url) is not None
+
+    @classmethod
+    def get_authors(cls, node):
+        authors = []
+        for person in node["dcterms:creator"]:
```
Comment (fako) [inline harvester/sources/extraction/edurep.py]: If we're going to talk to Edurep about the mime_type anyway. Let's ask about person identifiers.

File: harvester/sources/extraction/edurep.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                "email": None,
+                "external_id": None,
+                "dai": None,
+                "orcid": None,
+                "isni": None
+            })
+        return authors
+
+    @classmethod
+    def get_educational_levels(cls, node):
```
Comment (fako) [inline harvester/sources/extraction/edurep.py:147]: There are indeed different versions of what seems to be the same at first glance. The difference is that there are NL-LOM allowed and/or recommended values. These get extracted in the OAI-PMH version and are availble under lom_educational_levels, but are not available in JSON Search as fas as I know. And then there is this educational_level variant where Edurep stores values that they made up themselves. We need to properly clean this, but they are not offering the alternative anymore. So we'll stick to this one.

File: harvester/sources/extraction/edurep.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    @classmethod
+    def get_material_types(cls, node):
+        material_types = node["schema:learningResourceType"]
+        if not material_types:
+            return []
+        return [
+            material_type["schema:termCode"].text.strip()
+            for material_type in material_types
+        ]
+
```
Comment (fako) [inline harvester/sources/extraction/edurep.py:227]: Lint violations. Check the linter en tests.

File: harvester/sources/tests/extraction/test_edurep.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        super().setUpClass()
+        cls.begin_of_time = make_aware(datetime(year=1970, month=1, day=1))
+        EdurepJsonSearchResourceFactory.create_common_responses()
+        cls.seeds = get_harvest_seeds(Repositories.EDUREP_JSONSEARCH, SET_SPECIFICATION, cls.begin_of_time)
+
+    def test_get_id(self):
+        seeds = self.seeds
+        self.assertEqual(seeds[0]["external_id"], "7288bd68-d62b-4db0-8cea-5f189e209254")
+
+#     def test_get_files(self):
```
Comment (fako) [inline harvester/sources/tests/extraction/test_edurep.py]: More tests! :smile:

**Review by fako — CHANGES_REQUESTED** [review]
Nice start! Let's look at the disciplines/ideas/studies part together and write an email to Edurep with some questions to see if we can improve on the data. If we don't ask it will never improve! For the rest some minor things to change as well as tests to add. Also the comparison tests between OAI-PMH and JSON Search

File: harvester/sources/extraction/edurep.py (hunk 1 above)
Comment (kallepronk) [reply harvester/sources/extraction/edurep.py]: This has been adapted from another file, so far as I could find there is only a technical type.

File: harvester/sources/extraction/edurep.py (hunk 4 above)
Comment (kallepronk) [reply harvester/sources/extraction/edurep.py]: I've made a change to that method, does that change correspond to what you were thinking?

File: harvester/sources/extraction/buas.py [hunk 9]
```diff
@@ -215,6 +215,5 @@ def get_analysis_allowed(cls, node):
     "copyright_description": lambda node: None,
     "learning_material_disciplines": lambda node: [],
     "consortium": lambda node: None,
-    "lom_educational_level": lambda node: None,
```
Comment (fako) [inline harvester/sources/extraction/buas.py:230]: Beetje een misverstand misschien, maar als je dit niet bij alle bronnen gelijk gaat trekken, dan kun je het beter met rust laten. Het bij een enkele bron verwijderen zal niet goed gaan. Bij Sharekit wordt het nog gebruikt.

File: harvester/sources/extraction/edurep.py (hunk 1 above)
Comment (fako) [reply harvester/sources/extraction/edurep.py]: For the record ... that file was HKU and HKU's "API" is a bunch of JSON files on a server somewhere with inconsistent typing. We shouldn't lower the bar that far if possible.

File: harvester/sources/extraction/edurep.py [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+            date = node.get("schema:dateModified", None)
+        if date is None:
+            date = node.get("schema:publisherDate", None)
+        if date is None:
+            return None
+        return date
+
+    @classmethod
+    def get_publisher_year(cls, node):
+
```
Comment (fako) [inline harvester/sources/extraction/edurep.py:241]: Remove whitespace

File: harvester/sources/extraction/edurep.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+    @classmethod
+    def get_publisher(cls, node):
+        publisher = node.get("dcterms:publisher", None)
+        if publisher is None:
+            return "None"
+        else:
+            return publisher
+
+EDUREP_EXTRACTION_OBJECTIVE = {
+    # Essential NPPO properties
```
Comment (fako) [inline harvester/sources/extraction/edurep.py:270]: Edurep is not NPPO (aka Publinova), but Edusources. Maybe just remove all the comments. We'll separate Edusources/Publinova specific properties another day (in code not just comments)

**Review by fako — COMMENTED** [review]
Only small stuff. Look at Flake8. We'll fix the tests on Github together

**Pull request closed without merging**